paddle/fluid/operators/expand_v2_op_npu.cc

```diff
@@ -0,0 +1,184 @@
+
+    framework::DDim out_dims = framework::make_ddim(final_expand_shape);
+    Out->Resize(out_dims);
+    Out->mutable_data<T>(ctx.GetPlace());
+
+    const auto& runner = NpuOpRunner("ExpandD", {*X}, {*Out}, attr_input);
+    auto stream =
+        ctx.template device_context<paddle::platform::NPUDeviceContext>()
+            .stream();
+    runner.Run(stream);
+  }
+};
+
+template <typename DeviceContext, typename T>
+class ExpandV2NPUGradKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* dout = ctx.Input<Tensor>(framework::GradVarName("Out"));
+    auto* dx = ctx.Output<Tensor>(framework::GradVarName("X"));
+    dx->mutable_data<T>(ctx.GetPlace());
+
+    auto stream =
+        ctx.template device_context<paddle::platform::NPUDeviceContext>()
+            .stream();
+
+    // case 1: reduce dout dims to dx dims
+    // For example: [2, 120] --> [120]
+    auto reduce_ndim = dout->dims().size() - dx->dims().size();
+    std::vector<int> axes;
+    for (auto i = 0; i < reduce_ndim; ++i) {
+      axes.push_back(i);
+    }
+    Tensor* tmp_dout = const_cast<Tensor*>(dout);
+    Tensor reduced_dout(dx->type());
+    if (axes.size() != 0) {
+      std::vector<int64_t> reduced_dout_dims;
+      for (auto i = reduce_ndim; i < dout->dims().size(); ++i) {
+        reduced_dout_dims.push_back(dout->dims()[i]);
+      }
+      reduced_dout.Resize(framework::make_ddim(reduced_dout_dims));
+      reduced_dout.mutable_data<T>(ctx.GetPlace());
+      const auto& runner = NpuOpRunner("ReduceSumD", {*dout}, {reduced_dout},
+                                       {{"axes", axes}, {"keep_dims", false}});
+      runner.Run(stream);
+      tmp_dout = &reduced_dout;
+    }
+
+    // case 2: reduce axis of dout in which dim is 1
+    // For example: [12, 140] --> [1, 140]
+
+    // case 3: copy dout to dx when shape is totally same, and dim in dx != 1
+    // For example: [2, 10, 5] --> [2, 10, 5]
+    axes.clear();
+    for (auto i = 0; i < dx->dims().size(); ++i) {
+      if (dx->dims()[i] == 1) {
+        axes.push_back(i);
+      }
+    }
+    if (axes.size() != 0) {
+      const auto& runner = NpuOpRunner("ReduceSumD", {*tmp_dout}, {*dx},
+                                       {{"axes", axes}, {"keep_dims", true}});
+      runner.Run(stream);
+    } else {
+      framework::TensorCopySync(*tmp_dout, ctx.GetPlace(), dx);
+    }
+  }
+};
+
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+REGISTER_OP_NPU_KERNEL(
+    expand_v2,
+    ops::ExpandV2NPUKernel<paddle::platform::NPUDeviceContext, float>,
+    ops::ExpandV2NPUKernel<paddle::platform::NPUDeviceContext,
+                           paddle::platform::float16>,
+    ops::ExpandV2NPUKernel<paddle::platform::NPUDeviceContext, int>);
+
+REGISTER_OP_NPU_KERNEL(
+    expand_v2_grad,
+    ops::ExpandV2NPUGradKernel<paddle::platform::NPUDeviceContext, float>,
```

Comment:
Seems missing fp16 kernel  for expand_grad

Reply:
add fp16 kernel support for expand_grad